### PR TITLE
feat(channel): Add channel component

### PR DESCRIPTION
**Pull request by jdhughes-usgs**
Channel functionality limited to calculating lateral inflow to segments. Required
refactoring of netcdf functionality so that data that is dimensioned by nsegments
can be written. Added metadata to csv functionality so netcdf output would use
metadata for variables, if available. Added a few helper functions to metadata class.

### CODE REVIEW

**Review by jmccreight — CHANGES_REQUESTED** [review]
overall, a real peach! actually not being cynical. this is a nice piece of coding to read and its clarity is helpful at showing potential improvements to the existing scheme.

File: pynhm/base/meta.py [hunk 1]
```diff
@@ -14,6 +16,53 @@ def load_yaml_file(the_file):
...
+        raise ValueError(f"invalid metadata type '{type_str}'")
+    return netcdf_type_str
+
+
+def meta_numpy_type(meta_item: dict) -> str:
+    type_str = meta_type(meta_item)
+    if type_str == "I":
+        numpy_type_str = int
+    elif type_str == "B":
+        numpy_type_str = int
```
[the remark this answers is not in the document]
Comment (jmccreight) [reply pynhm/base/meta.py:39]: might postpone

File: pynhm/base/meta.py [hunk 2]
```diff
@@ -62,3 +123,49 @@ def get_params(self, inputs_list: Iterable) -> dict:
...
+        """
+        if isinstance(variables, str):
+            variables = [variables]
+        variable_dict = self.find_variables(variables)
+        return {
+            key: [dimension for tag, dimension in value["dimensions"].items()]
+            for key, value in variable_dict.items()
+        }
+
+    def find_variables(self, variables: varoptions) -> dict:
```
[the remark this answers is not in the document]
Comment (jdhughes-usgs) [reply pynhm/base/meta.py:150]: Added description to docstring - function ok

File: pynhm/utils/netcdf_utils.py [hunk 3]
```diff
@@ -204,41 +251,81 @@ def __init__(
...
-            nhrus = len(hru_ids)
-        elif isinstance(hru_ids, np.ndarray):
-            nhrus = hru_ids.shape[0]
-        self.nhrus = nhrus
-        self.hru_ids = hru_ids
+        variable_dimensions = {}
+        nhru_coordinate = False
+        nsegment_coordinate = False
+        for var_name in variables:
+            dimension_name = meta_dimensions(var_meta[var_name])
```
[the remark this answers is not in the document]
Comment (jdhughes-usgs) [reply pynhm/utils/netcdf_utils.py:258]: Consider further later.

File: pynhm/utils/netcdf_utils.py [hunk 4]
```diff
@@ -204,41 +251,81 @@ def __init__(
...
+            variable_dimensions[var_name] = dimension_name
+            if "nhru" in dimension_name or "ngw" in dimension_name:
+                nhru_coordinate = True
+            if "nsegment" in dimension_name:
+                nsegment_coordinate = True
+
+        if nhru_coordinate:
+            hru_ids = coordinates["nhm_id"]
```
[the remark this answers is not in the document]
Comment (jdhughes-usgs) [reply pynhm/utils/netcdf_utils.py:266]: Table for now

File: pynhm/hydrology/PRMSChannel.py [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+        else:
+            segment_order = [0]
+        self._segment_order = np.array(segment_order, dtype=int)
+
+        # calculate the Muskingum parameters
+        velocity = (
+            (
+                (1.0 / self.mann_n)
+                * np.sqrt(self.seg_slope)
+                * self.seg_depth ** (2.0 / 3.0)
```
[the remark this answers is not in the document]
Comment (jdhughes-usgs) [reply pynhm/hydrology/PRMSChannel.py:98]: that is correct.

File: autotest/test_channel.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        input_variables = {}
+        for key in PRMSChannel.get_inputs():
+            nc_path = output_dir / f"{key}.nc"
+            input_variables[key] = nc_path
+
+        channel = PRMSChannel(control, prms_params, **input_variables)
+        nc_parent = tmp_path / domain["domain_name"]
+        channel.initialize_netcdf(nc_parent)
+
+        output_compare = {}
```
[the remark this answers is not in the document]
Comment (jdhughes-usgs) [reply autotest/test_channel.py:30]: Moved what could be moved

File: autotest/test_preprocess_csv.py [hunk 7]
```diff
@@ -14,7 +14,8 @@ def compare_netcdf(csv, nc_name):
     ds = nc.Dataset(nc_name)
 
     np_data = csv.data
-    hrus = csv.hru_ids
+    hrus = csv.nhm_id
+    _ = csv.nhm_seg
```
[the remark this answers is not in the document]
Comment (jdhughes-usgs) [reply autotest/test_preprocess_csv.py:18]: It exists

File: pynhm/base/meta.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jdhughes-usgs) [reply pynhm/base/meta.py:39]: Leave for later

File: pynhm/base/meta.py [hunk 8]
```diff
@@ -14,6 +16,53 @@ def load_yaml_file(the_file):
...
+
+    Args:
+        meta_item: meta dictionary for a variable name
+
+    Returns:
+        dimensions: tuple with dimension strings for a variable name
+
+    """
+    return [value for key, value in meta_item["dimensions"].items()]
+
```
[the remark this answers is not in the document]
Comment (jdhughes-usgs) [reply pynhm/base/meta.py:64]: Done

File: pynhm/hydrology/PRMSChannel.py [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+    outflow_ts: np.ndarray,
+    seg_current_sum: np.ndarray,
+    tsi: np.ndarray,
+    ts: np.ndarray,
+    c0: np.ndarray,
+    c1: np.ndarray,
+    c2: np.ndarray,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Muskinggum routing function that calculates the upstream inflow and
```
[the remark this answers is not in the document]
Comment (jdhughes-usgs) [reply pynhm/hydrology/PRMSChannel.py:343]: No changes required. Assessment of what the muskingum_mann method does.

File: pynhm/hydrology/PRMSChannel.py [hunk 10]
```diff
@@ -0,0 +1,448 @@
...
+
+                    # todo: evaluated if denormal results should be dealt with
+
+                    # Muskingum routing equation
+                    outflow_ts[jseg] = (
+                        inflow_ts[jseg] * c0[jseg]
+                        + seg_inflow0[jseg] * c1[jseg]
+                        + outflow_ts[jseg] * c2[jseg]
+                    )
+                else:
```
[the remark this answers is not in the document]
Comment (jdhughes-usgs) [reply pynhm/hydrology/PRMSChannel.py:410]: No changes required. Assessment of what the muskingum_mann method does.

File: pynhm/hydrology/PRMSChannel.py [hunk 11]
```diff
@@ -0,0 +1,448 @@
...
+            # current inflow to the segment is the time-weighted average
+            # of the outflow of the upstream segments and the lateral HRU
+            # inflow plus any gains
+            seg_current_inflow = seg_lateral_inflow[jseg]
+
+            # todo: evaluate if obsin_segment needs to be implemented -
+            #  would be needed needed if headwater basins are not included
+            #  in a simulation
+
+            seg_current_inflow += seg_upstream_inflow[jseg]
```
[the remark this answers is not in the document]
Comment (jdhughes-usgs) [reply pynhm/hydrology/PRMSChannel.py:390]: No changes required. Assessment of what the muskingum_mann method does.

File: pynhm/hydrology/PRMSChannel.py [hunk 12]
```diff
@@ -0,0 +1,448 @@
...
+adaptable = Union[str, np.ndarray, Adapter]
+
+FT2_PER_ACRE = 43560.0
+INCHES_PER_FOOT = 12.0
+SECS_PER_HOUR = 3600.0
+
+
+class PRMSChannel(StorageUnit):
+    """PRMS channel flow
+
```
[the remark this answers is not in the document]
Comment (jdhughes-usgs) [reply pynhm/hydrology/PRMSChannel.py:23]: Added a summary of the method.

File: pynhm/hydrology/PRMSChannel.py [hunk 13]
```diff
@@ -0,0 +1,448 @@
...
+        # this is different from PRMS, which relied on divide by zero resulting
+        # in a value of infinity that when evaluated relative to a maximum
+        # desired Kcoef value of 24 would be reset to 24. This approach is
+        # equivalent and avoids the occurence of a divide by zero.
+        Kcoef = np.full(self.nsegment, 24.0, dtype=float)
+
+        # only calculate Kcoef for cells with velocities greater than zero
+        idx = velocity > 0.0
+        Kcoef[idx] = self.seg_length[idx] / velocity[idx]
+        Kcoef = np.where(self.segment_type == SegmentType.LAKE, 24.0, Kcoef)
```
[the remark this answers is not in the document]
Comment (jdhughes-usgs) [reply pynhm/hydrology/PRMSChannel.py:117]: Modified

File: pynhm/hydrology/PRMSChannel.py [hunk 14]
```diff
@@ -0,0 +1,448 @@
...
+        # local flow variables
+        self._seg_inflow = np.zeros(self.nsegment, dtype=float)
+        self._seg_inflow0 = np.zeros(self.nsegment, dtype=float)
+        self._seg_outflow0 = np.zeros(self.nsegment, dtype=float)
+        self._inflow_ts = np.zeros(self.nsegment, dtype=float)
+        self._outflow_ts = np.zeros(self.nsegment, dtype=float)
+        self._seg_current_sum = np.zeros(self.nsegment, dtype=float)
+
+        # initialize variables
+        self.seg_outflow = self.segment_flow_init
```
[the remark this answers is not in the document]
Comment (jdhughes-usgs) [reply pynhm/hydrology/PRMSChannel.py:181]: Moved

File: pynhm/hydrology/PRMSChannel.py [hunk 15]
```diff
@@ -0,0 +1,448 @@
...
+        idx = velocity > 0.0
+        Kcoef[idx] = self.seg_length[idx] / velocity[idx]
+        Kcoef = np.where(self.segment_type == SegmentType.LAKE, 24.0, Kcoef)
+        Kcoef = np.where(Kcoef < 0.01, 0.01, Kcoef)
+        self._Kcoef = np.where(Kcoef > 24.0, 24.0, Kcoef)
+
+        self._ts = np.ones(self.nsegment, dtype=float)
+        self._tsi = np.ones(self.nsegment, dtype=int)
+
+        # todo: vectorize this
```
[the remark this answers is not in the document]
Comment (jdhughes-usgs) [reply pynhm/hydrology/PRMSChannel.py:124]: No changes required. Assessment of what the muskingum_mann method does.

File: test_data/scripts/test_nc_domains.py [hunk 16]
```diff
@@ -1,9 +1,9 @@
-from datetime import timedelta
-import numpy as np
 import os
 import pathlib as pl
 import sys
+from datetime import timedelta
 
+import numpy as np
```
[the remark this answers is not in the document]
Comment (jdhughes-usgs) [reply test_data/scripts/test_nc_domains.py:6]: Revised imports

**Pull request merged**